sql-plugin/src/main/scala/com/nvidia/spark/rapids/BatchWithPartitionData.scala

```diff
@@ -0,0 +1,483 @@
+    val rowCountsForEachBatch = listOfPartitionedRowsData.map(partitionData =>
+      partitionData.map {
+        case PartitionRowData(_, rowNum) => rowNum
+      }.sum
+    )
+    // Calculate split indices using cumulative sum
+    rowCountsForEachBatch.scanLeft(0)(_ + _).drop(1).dropRight(1)
+  }
+
+  /**
+   * Split a ColumnarBatch into multiple ColumnarBatches based on indices.
+   */
+  private def splitColumnarBatch(input: ColumnarBatch,
+      splitIndices: Seq[Int]): Seq[SpillableColumnarBatch] = {
+    if (splitIndices.isEmpty) {
+      Seq(SpillableColumnarBatch(GpuColumnVector.incRefCounts(input),
+        SpillPriorities.ACTIVE_ON_DECK_PRIORITY))
+    } else {
+      val schema = GpuColumnVector.extractTypes(input)
+      val splitInput = withResource(GpuColumnVector.from(input)) { table =>
+        table.contiguousSplit(splitIndices: _*)
+      }
+      splitInput.safeMap { ct =>
+        SpillableColumnarBatch(ct, schema, SpillPriorities.ACTIVE_BATCHING_PRIORITY)
+      }
+    }
+  }
+
+  /**
+   * Splits an array of `PartitionRowData` into two halves of equal size.
+   *
+   * Example 1,
+   * {{{
+   * Input:
+   *   partitionedRowsData: [ (1000, [ab, cd]), (2000, [def, add]) ]
+   *   target rows: 1500
+   *
+   * Result:
+   *   [
+   *      [ (1000, [ab, cd]), (500 [def, add]) ],
+   *      [ (1500, [def, add]) ]
+   *   ]
+   * }}}
+   *
+   * Example 2,
+   * {{{
+   * Input:
+   *   partitionedRowsData: [ (1000, [ab, cd]) ]
+   *   target rows: 500
+   *
+   * Result:
+   *   [
+   *      [ (500, [ab, cd]) ],
+   *      [ (500, [ab, cd]) ]
+   *   ]
+   * }}}
+   *
+   * @note This function ensures that splitting is possible even in cases where there is a single
+   * large partition until there is only one row.
+   */
+  private def splitPartitionDataInHalf(
+      partitionedRowsData: Array[PartitionRowData]): Array[Array[PartitionRowData]] = {
+    val totalRows = partitionedRowsData.map(_.rowNum).sum
+    if (totalRows <= 1) {
+      throw new SplitAndRetryOOM("GPU OutOfMemory: cannot not split input with one row")
+    }
+    var remainingRows = totalRows / 2
+    val leftHalf = ArrayBuffer[PartitionRowData]()
+    val rightHalf = ArrayBuffer[PartitionRowData]()
+    partitionedRowsData.foreach { partitionRow: PartitionRowData =>
+      if (remainingRows > 0) {
+        // Determine how many rows to add to the left partition
+        val rowsToAddToLeft = Math.min(partitionRow.rowNum, remainingRows)
+        leftHalf.append(PartitionRowData(partitionRow.rowValue, rowsToAddToLeft))
+        remainingRows -= rowsToAddToLeft
+        if (remainingRows <= 0) {
+          // Add remaining rows to the right partition
+          rightHalf.append(PartitionRowData(partitionRow.rowValue,
+            partitionRow.rowNum - rowsToAddToLeft))
+        }
+      } else {
+        rightHalf.append(partitionRow)
+      }
+    }
+    val totalRowsInSplits = leftHalf.map(_.rowNum).sum + rightHalf.map(_.rowNum).sum
```

Comment:
Nit: It's a bit odd to traverse the data twice here, as we could trivially track the total number of rows processed as we're looping through.